### DISCUSSION TITLE
Custom model with predict_proba(X) method - should models.evaluate work with that?

### DISCUSSION THREAD

**Discussion by JU102030** [mlflow/mlflow#18754] [Q&A]
**In short**: is `mlflow.models.evaluate(..)` meant to detect and use `detect_proba(X)` methods on custom classifier models and thus produce the ROC / Lift / Precision-Recall artefacts and related metrics?

I don't think this works (MLFlow 3.5.1) and I am not sure whether I am doing something wrong, or if this is a feature request or a bug?

**Longer Question**:
I have a custom PyFunc model roughly as follows:

```
class DistEstModel(mlflow.pyfunc.PythonModel):
    
    def predict_proba(self, model_input):
        ... some custom code return a (-1,2) shaped NumPy tensor...

    def predict(self, model_input, params=None):
         .. some custom code which uses predict_proba. Returns a (-1,) shaped Int tensor with classifications ...

    def load_context(self, context):
        return super().load_context(context)
    
    def fit(self, X_train, y_train, **kwargs):
        ... more code ...

```
The model is a classifier returning an int64 numpy array with 1 for the positive case and 0 for the negative case. `predict_proba` returns a (-1,2) shaped Float NumPy Array with probabilities for the "0" and "1" cases (adding up to 1).

The model is instantiated with `dem = DistEstModel()` and fit is called. 

I then register and evaluate the model as follows (The various X/y test/train variables are all NumPy arrays. `y_pred_dem` are the trained model predictions.)

```
with mlflow.start_run(run_name="kde-eval-test") as run:
    
    dataset = mlflow.data.from_numpy(features=X_train, targets=y_train, name="scikit-autogen") 
    mlflow.log_input(dataset, context="training")
    
    signature = mlflow.models.infer_signature(X_test, y_pred_dem)

    model_info = mlflow.pyfunc.log_model(
        name=dem._modelname,
        python_model=dem,
        input_example=X_test,
        tags=dem.tags,
        signature=signature,
        model_type="classifier"
    )

    test_dataset = mlflow.data.from_numpy(
        features=X_test, 
        targets=y_test,
        name="scikit-autogen-test", 
    )
    mlflow.log_input(test_dataset, context="training-eval")

    result = mlflow.models.evaluate(
        model=model_info.model_uri,
        data=test_dataset,
        model_type="classifier",
        evaluators=["default"]
    )

```

The model works fine both in an mlflow API served deployment as well as in a batch use case in a completely different project / python environment which retrieves it from the MLFlow server - (in other words I can be sure that `predict_proba` has been serialised correctly and is present.

The issue / question I have is this: The `evaluate` call only produces the confusion matrix and Shap artefacts, but not ROC, recall/precision and lift. It also doesn't produce the _auc metric. (Other classification metrics such as accuracy, recall, f1 ... are produced)

I dug into the MLFlow code and I know why this happens:
* in `default_evaluator.py` `_extract_raw_model` returns `None` for its second returned tuple element - i.e. it fails to find/extract the underlying model class (presumably `DistEstModel` in my case). Looking at the code it seems its only meant to function for catboost/xgboost/sklearn models anyway and would never work for custom models. 
* in `models/evaluation/evaluators/classifier.py` in `_extract_predict_fn_and_prodict_proba_fn` this then means that the `predict_proba` function is never recognised.
* As a consequence of that, `predict_proba` is never called and in subsequent code `y_prob` ends up as `None` which then ultimately means that ROC and anything else that needs it for evaluation doesn't work.

My questions is:
- I am doing something wrong? (In other words: What is the correct way to make customer models compliant with `evaluate(...)` in case of classifiers?
- Is this a bug?
- Is this a (sensible) feature request?

Happy to raise the bug/feature if people think this is appropriate.

Kind Regards
Jens

**Comment by JU102030** [comment]
Actually, I just found [FR 5337](https://github.com/mlflow/mlflow/issues/5337) which seems to request exactly what I am asking for - making `.evaluate` work with custom models `predict_proba`.  
That FR has been open for a while - if I was to implement this, would people be interested? would this (likely) be reviewed and merged?

**Comment by aayostem** [comment]
@JU102030  The limitation is indeed in MLflow's evaluator, which currently only automatically detects predict_proba for specific supported frameworks (scikit-learn, XGBoost, LightGBM, etc.).
I think this is a feature gap rather than a bug. The current implementation intentionally focuses on major ML frameworks.
But as you discovered, FR #5337 exactly addresses this limitation. 

However, you can  manually evaluate with your predict_proba
```
# Use your model's predict_proba directly for manual evaluation
model = mlflow.pyfunc.load_model(model_info.model_uri)
y_proba = model.predict_proba(X_test)
y_pred = model.predict(X_test)

# Calculate metrics manually using sklearn
from sklearn.metrics import roc_auc_score, roc_curve, precision_recall_curve
auc = roc_auc_score(y_test, y_proba[:, 1])
fpr, tpr, _ = roc_curve(y_test, y_proba[:, 1])

# Log these manually to MLflow
mlflow.log_metric("roc_auc", auc)
# ... log other metrics and plots
```

i hope this helps you

**Comment by JU102030** [comment]
Thanks! - appreciate the response! I knew I could go manual, I just figured it would be convenient if the automation also covers these cases. There isn't a whole lot of activity on FR#5337 -  lets see if that eventually makes it.
